pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,166 @@
+  }
+
+  /**
+   * @param i 0-indexed position in the order of first appearance
+   * @return The name of the placeholder at this backend parameter position
+   */
+  public String getPlaceholderName(int i) {
+    if (placeholderNames == null) {
+      throw new RuntimeException("Call hasNamedParameters() first.");
+    }
+    return placeholderNames.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the SQL text for this placeholder index
+   */
+  public int getPlaceholderPosition(int i) {
+    if (placeholderPositions == null) {
+      throw new RuntimeException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the order of first appearance of each placeholder
+   */
+  public int getPlaceholderAtPosition(int i) {
+    if (placeholderAtPosition == null) {
+      throw new RuntimeException("Call hasParameters() first.");
+    }
+    return placeholderAtPosition.get(i);
+  }
+
+  public int getLatestPlaceholderPosition() {
+    if (placeholderPositions == null) {
+      throw new RuntimeException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(placeholderPositions.size() - 1);
+  }
+
+  public int addNamedParameter(int position, String bindName) throws SQLException {
+    if (bindStyle == BindStyle.POSITIONAL) {
+      throw new SQLException("Positional and named parameters cannot be combined! Saw "
+          + "positional parameter first.");
+    }
+
+    if (bindStyle == null) {
+      bindStyle = BindStyle.NAMED;
+    }
+
+    if (placeholderPositions == null) {
+      placeholderPositions = new ArrayList<>();
+    }
+    placeholderPositions.add(position);
+
+    if (placeholderNames == null) {
+      placeholderNames = new ArrayList<>();
+    }
+    int bindIndex = placeholderNames.indexOf(bindName);
+    if (bindIndex == -1) {
+      bindIndex = placeholderNames.size();
+      placeholderNames.add(bindName);
+    }
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+    placeholderAtPosition.add(bindIndex);
+    return bindIndex + 1;
+  }
+
+  /**
+   * @return Returns the number of placeholder appearances in the SQL text
+   */
+  public int placeholderCount() {
+    return placeholderPositions == null ? 0 : placeholderPositions.size();
+  }
+
+  /**
+   * @return Returns the number of parameter to be sent to the backend.
+   */
+  public int nativeParameterCount() {
+    if (placeholderNames != null) {
+      return placeholderNames.size();
+    }
+
+    return placeholderCount();
+  }
+
+  public List<Integer> getPlaceholderPositions() {
+    if (placeholderPositions == null) {
+      return NO_PLACEHOLDERS;
+    }
+    return placeholderPositions;
+  }
+
+  enum BindStyle {
```

Comment:
@davecramer, @vlsi does this project have common standard on where nested enums should be declared? I am more accustomed to seeing at beginning of class, but this is purely style/convention.

Reply:
I'd prefer they were at the top as well. We don't have a standard, but I agree with putting them at the top

Reply:
That is my preference as well, but the Intellij formatter configuration in the project wants it at the bottom. Fixed.